### PR TITLE
Add support for sections

### PR DESCRIPTION
**Pull request by riccardoferretti**
This PR adds support for sections in Foam.

That means that:
- `[[resource#section]]` will be recognized as a valid resource and not as a placeholder
- `[[resource#section]]` will navigate to the `section` line in the `resource` 
- Autocompletion for sections
- `![[resource#section]]` will embed only the text in `section` in the preview
- Diagnostics will report the usage of incorrect sections (e.g. `[[resource#section3]]` when no `section 3` exists in the file), with quick suggestions to fix the error

Note for reviewers: Even after the PR has been merged, feel free to leave comments, I will address them in a follow-up PR.  This is to not slow down work while waiting for feedback.

### CODE REVIEW

File: packages/foam-vscode/src/core/model/note.ts [hunk 1]
```diff
@@ -38,12 +38,17 @@ export interface Tag {
...
+  range: Range;
+}
+
 export interface Resource {
   uri: URI;
   type: string;
   title: string;
   properties: any;
-  // sections: NoteSection[]
+  blocks: NoteBlock[];
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/core/model/note.ts:51]: Calling this `blocks` so that in the future it will also house inline blocks (https://github.com/foambubble/foam/issues/294) - sections are just a type of block

File: packages/foam-vscode/src/features/utility-commands.ts [hunk 2]
```diff
@@ -1,54 +1,72 @@
...
 };
 
 const feature: FoamFeature = {
-  activate: (context: vscode.ExtensionContext) => {
+  activate: (context: vscode.ExtensionContext, foamPromise: Promise<Foam>) => {
     context.subscriptions.push(
       vscode.commands.registerCommand(
         OPEN_COMMAND.command,
-        OPEN_COMMAND.execute
+        async (params: { uri: URI }) => {
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/features/utility-commands.ts:23]: moved the `execute` here as we need `foam` here and trying to be smart about initialization was more convoluted than just doing this.

Commands need a bit of love/refactoring, but this is out of scope for this PR

File: packages/foam-vscode/src/core/markdown-provider.ts [hunk 3]
```diff
@@ -201,6 +222,53 @@ const tagsPlugin: ParserPlugin = {
   },
 };
 
+let blockStack: Array<{ label: string; level: number; start: Position }> = [];
+const blocksPlugin: ParserPlugin = {
```
Comment (memeplex) [inline packages/foam-vscode/src/core/markdown-provider.ts:226]: Not specific of this change, but how well does reparsing the entire file on each change scale? Do you know? I see it's an async operation, but anyway it could be pretty expensive for big files. In the ZK many-small-files scenario is not problem, but if you have some big files in the same workspace, especially now that there is section linking support, couldn't it be a problem?

Comment (riccardoferretti) [reply packages/foam-vscode/src/core/markdown-provider.ts:226]: Performance is something that I have been postponing. At some point we'll need to take a serious look at it (starting from whether using a parser is better than regexes, which other tools have preferred). But I have (semi-)consciously decided to leave it for later

File: packages/foam-vscode/src/core/markdown-provider.ts [hunk 4]
```diff
@@ -201,6 +222,53 @@ const tagsPlugin: ParserPlugin = {
...
+  name: 'block',
+  onWillVisitTree: () => {
+    blockStack = [];
+  },
+  visit: (node, note) => {
+    if (node.type === 'heading') {
+      const level = (node as any).depth;
+      const label = ((node as Parent)!.children?.[0] as any)?.value;
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/markdown-provider.ts:234]: Not sure I follow, how would you change the line? 
I admit sometimes I get lost in TS's types so I have overdone it haha

[the remark this answers is not in the document]
Comment (memeplex) [reply packages/foam-vscode/src/core/markdown-provider.ts:234]: (anyway it's done this way in other places so it's probably better to leave it alone just for the sake of consistency, but would you mind explaining it?)

File: packages/foam-vscode/src/core/model/note.ts (hunk 1 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/note.ts:51]: I see your point about blocks, it is inspired by names we have used before, but I also don't love it.
I am not sure about Section as I don't think a section represents well an "inline block".. mm, I am torn about it.
It's almost an Anchor, but that is a point, not a range.
I guess it could also be considered a Fragment, but that feels confusing with `URI.fragment`.

Other ideas?

File: packages/foam-vscode/src/core/model/workspace.ts [hunk 5]
```diff
@@ -152,11 +165,23 @@ export class FoamWorkspace implements IDisposable {
...
+    if (!resource) {
+      return null;
+    }
+    if (!fragment) {
+      return resource;
+    }
+    return {
+      ...resource,
+      uri: URI.withFragment(resource.uri, fragment),
+    };
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/workspace.ts:184]: Yes, my thinking here was that if the section doesn't exist we simply return the whole thing. 
We fallback gracefully, returning `null` feels too strict to me.
That being said, there might be some unexpected consequences, which might require us to reconsider this in the future

File: packages/foam-vscode/src/features/link-completion.ts [hunk 6]
```diff
@@ -19,11 +18,64 @@ const feature: FoamFeature = {
         mdDocSelector,
         new CompletionProvider(foam.workspace, foam.graph),
         '['
+      ),
+      vscode.languages.registerCompletionItemProvider(
+        mdDocSelector,
+        new SectionCompletionProvider(foam.workspace, foam.graph),
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/link-completion.ts:24]: Ooops, good point. This needs to be consolidated once we figure out the most appropriate name for the field/concept

File: packages/foam-vscode/src/features/navigation-provider.ts [hunk 7]
```diff
@@ -110,20 +110,22 @@ export class NavigationProvider
...
-        )
-      ),
+      targetRange: toVsCodeRange(targetRange),
       targetSelectionRange: toVsCodeRange(
-        Range.createFromPosition(
-          targetResource.source.contentStart,
-          targetResource.source.contentStart
-        )
+        Range.createFromPosition(targetRange.start, targetRange.start)
       ),
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/navigation-provider.ts:129]: `targetSelectionRange` with same start and end is so is so that there will be no selection in the editor when navigating the link

File: packages/foam-vscode/src/core/model/workspace.ts [hunk 8]
```diff
@@ -152,11 +165,23 @@ export class FoamWorkspace implements IDisposable {
...
+
+    if (!resource) {
+      return null;
+    }
+    if (!fragment) {
+      return resource;
+    }
+    return {
+      ...resource,
+      uri: URI.withFragment(resource.uri, fragment),
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/workspace.ts:183]: Yes, I see your point, and this is related to the comment below.

> the representation of a section is a resource that is identical to its entire note except that the uri includes a fragment? That is, all sections in a note have the same representation except for the different fragments in their uris

That's correct.

> may be worth encapsulating as part of the Resource abstraction

I think that's a very good idea

**Review by riccardoferretti — COMMENTED** [review]
Thanks for the comments, some good pointers in there. 

I am mainly thinking about the Block vs Section debate, I see your point, and the more I think about it (even from when I wrote my reply), the more I tend to agree with you.
Let me think about that one

File: packages/foam-vscode/src/core/markdown-provider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (memeplex) [reply packages/foam-vscode/src/core/markdown-provider.ts:234]: I'm just learning about the type system as we speak but I think that once you say `x is T` you don't need to also signal the compiler it can't be null by adding the `!`. Consider this example:

```ts
interface InnerThing {
    realThing: number;
}

interface Thing {
    innerThing?: InnerThing;
}

function doStuffWith(aThing: Thing, anotherThing: Thing) : number {
    return (aThing.innerThing as InnerThing).realThing + anotherThing.innerThing!.realThing;
}
```

When  compiled with strict null checking the last line can't just be  `aThing.innerThing.realThing + anotherThing.innerThing.realThing`, but  both  doing the cast  or adding the `!` would be enough to silence the compiler, you don't need to do both of them.

That said, it doesn't do any harm and since it's consistently used in a similar way at other locations maybe it's better not to touch it.

**Comment by memeplex** [comment]
A couple other things:
1. How will this work with the graph? As it is now, are sections shown at all? What kind of relationship they have with their containing node?
2. What about links to a section in the same file? Suppose I type `[[#`, will that trigger a completion? Will foam be able to solve a link `#...` to a section in its containing file?

**Comment by memeplex** [comment]
Also:

3. What about backlinks?

By looking at  the code I see that they are now based on paths, not uris, so backlinks linking to the file or any of its sections will be shown in the backlink view. Another possibility is to dynamically update this view according to the current section (the section where the cursor is). The second possibility is more sophisticated and is what org-roam does, but Obsidian implements the first one and, since it's simpler and AFAICS (without actually running it) it should be already working in Foam, perhaps it's better to leave it like that. It could be a problem for very big files with many linked sections but I don't see that as a recommended usage pattern for Foam.

File: packages/foam-vscode/src/core/model/workspace.ts [hunk 9]
```diff
@@ -112,36 +112,49 @@ export class FoamWorkspace implements IDisposable {
...
-        const uri = resourceId as URI;
-        return this.exists(uri)
-          ? this.resources.get(normalize(uri.path)) ?? null
-          : null;
+    if (refType === 'uri') {
+      const uri = resourceId as URI;
+      return this.exists(uri)
+        ? this.resources.get(normalize(uri.path)) ?? null
+        : null;
+    }
```
Comment (memeplex) [inline packages/foam-vscode/src/core/model/workspace.ts:138]: I believe you can reorganize the code in this function in a different way:

```ts
switch (refType) {
      case 'key':
           ...
           return sorted[0];
      case 'absolute-path':
           ...
           resourceUri = ...;
           break;
      case 'relative-path':
           ...
           resourceUri = ...;
           break;
      case 'uri':
           resourceUri = resourceId;
           break;
}
return this.exists(resourceUri)
        ? this.resources.get(normalize(resourceUri.path)) ?? null
        : null;
```

This:
- Makes it clearer that ultimately what is resolved in the workspace is a uri.
- Deduplicates some normalization & resolution logic.
- Deals with the confusing fact of having three names (`uri`, `resourceUri` and `targetUri`) for what is essentially the same thing.

Unfortunately the `key` case has to be dealt in a different way because `listById` already returns resources, but all in all I think this will be an improvement.

**Comment by riccardoferretti** [comment]
> How will this work with the graph? As it is now, are sections shown at all? What kind of relationship they have with their containing node?

The graph only highlights connections amongst resources, it won't go into the detail of the section

> What about links to a section in the same file? Suppose I type [[#, will that trigger a completion? Will foam be able to solve a link #... to a section in its containing file?

Currently it doesn't but that's a good point, I will add that.

> What about backlinks?

Like the graph, backlinks are based on connections amongst files, not sections. I don't believe it's worth going down that level of detail, and it could actually be unintuitive and hard to convey to the user.

File: packages/foam-vscode/src/core/model/workspace.ts (hunk 9 above)
Comment (memeplex) [reply packages/foam-vscode/src/core/model/workspace.ts:138]: Also the code:

```ts
      return this.exists(uri)
        ? this.resources.get(normalize(uri.path)) ?? null
        : null;
```

in the `uri` clause seems a bit redundant: `exists` checks if the uri is a placeholder and then if `this.resources.get(normalize(uri.path))` is not null. So here you can just check if the uri is a placeholder, since `this.resources.get(normalize(uri.path))` will return null anyway when not found. This way you avoid the double resolution. Moreover, I don't believe the `?? null` is adding anything.

Comment (memeplex) [reply packages/foam-vscode/src/core/model/workspace.ts:138]: Indeed `exists` is only used here and doesn't add too much to the API. After the change above, you can remove `exists` since it's easily replaced by `isSome(ws.find(uri))`, or keep it in the `API` just like an unchecked version of `ws.get(uri)`:

```ts
  public exists(uri: URI): boolean {  // <-- doesn't add too much, maybe remove it
    return isSome(this.find(uri));
  }

  public get(uri: URI): Resource {
    const note = this.find(uri);
    if (isSome(note)) {
      return note;
    } else {
      throw new Error('Resource not found: ' + uri.path);
    }
  }
```

File: packages/foam-vscode/src/core/model/note.ts (hunk 1 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/note.ts]: decided to go for "section"

File: packages/foam-vscode/src/core/model/workspace.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/workspace.ts:183]: done

File: packages/foam-vscode/src/core/model/workspace.ts (hunk 9 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/core/model/workspace.ts:138]: Good pointers, I have implemented some in the PR
